### PR TITLE
[ARCTIC-941]Improve unit test cases in module core

### LINKED ISSUE #941: [Subtask]: Improve unit test for module core
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/arctic/issues?q=is%3Aissue) and found no similar issues.

### Description

The unit test quality of the core module will be improved through the following measures：

- add checkstyle for test codes
- add more TestBase implementation for writing test codes
- add more TestHelpers implementation for writing test codes
- complete unit test cases to cover a wider range of scenarios
- add a unit test code example to guide writing test codes

### Parent issue

#937

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by zhoujinsong**
## Why are the changes needed?

*To resolve #941, improve unit test cases in module core and hive*

## Brief change log

  - *Add TestAms and TestHMS external resource implementation*
  - *Add CatalogTestBase and TableTestBase implementation*
  - *Improve test cases and add tests for exception scenarios*

## How was this patch tested?
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

File: hive/src/test/java/com/netease/arctic/hive/catalog/HiveCatalogLoaderTest.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.Map;
+
+public class HiveCatalogLoaderTest {
+  @ClassRule
+  public static TestAms TEST_AMS = new TestAms();
+
+  @ClassRule
+  public static TestHMS TEST_HMS = new TestHMS();
+
+  private static final String TEST_CATALOG_NAME = "test";
```
Comment (baiyangtx) [inline hive/src/test/java/com/netease/arctic/hive/catalog/HiveCatalogLoaderTest.java:43]: if TEST_HMS is a @ClassRule, the HMS will be created and destroyed at each test class.

Comment (zhoujinsong) [reply hive/src/test/java/com/netease/arctic/hive/catalog/HiveCatalogLoaderTest.java:43]: I add a new system property `singleton-test-resource` to support only creating one test resource for all test cases. You can find related codes in class `SingletonResourceUtil` 、`TestAms` and `TestHMS`.
I plan to set `singleton-test-resource` to `true` in our Github CI execution to reduce resource creating cost.

File: site/docs/ch/roadmap.md [hunk 2]
```diff
@@ -11,7 +11,7 @@ Arctic 预计每2-3个月发布一个大版本，每1个月发布一个小版本
 | [0.4.0](https://github.com/NetEase/arctic/milestone/3)  | 2022.11          |
 | [0.5.0](https://github.com/NetEase/arctic/milestone/4)  | 2022.12          |
 
-## 开发中
+## 开发中T
```
Comment (baiyangtx) [inline site/docs/ch/roadmap.md:14]: T?

**Review by baiyangtx — COMMENTED** [review]
New test case classes are added to ArcticHiveTestMain. so there will be 2 HMS created during tests,

File: core/src/test/java/com/netease/arctic/TestResourceBase.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netease.arctic;
+
+public class TestResourceBase {
+}
```
[the remark this answers is not in the document]
Comment (zhoujinsong) [reply core/src/test/java/com/netease/arctic/TestResourceBase.java]: removed.

**Comment by YesOrNo828** [comment]
@zhoujinsong  Maybe we should consider abstracting the common logic to make HMS/Kafka Service/AMS resources to be singleton resources.

File: core/src/test/java/com/netease/arctic/TestAms.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      }
+    } else {
+      mockAms.start();
+      LOG.info("Start mock AMS before testing.");
+    }
+  }
+
+  @Override
+  protected void after() {
+    if (!SingletonResourceUtil.isUseSingletonResource()) {
```
[the remark this answers is not in the document]
Comment (zhoujinsong) [reply core/src/test/java/com/netease/arctic/TestAms.java:74]: For now, a shutdown hook is added to clean the singleton resource.

**Comment by zhoujinsong** [comment]
> @zhoujinsong Maybe we should consider abstracting the common logic to make HMS/Kafka Service/AMS resources to be singleton resources.

The singleton resource logic is static code now, it seems to be hard to abstract this logic.
Do you have a good idea about that?

**Comment by YesOrNo828** [comment]
> 
@zhoujinsong  Below HMS service demo code abstract singleton test service. 

```java
public class HiveBase {

  public static void setup() {
    System.out.println("================== setup HiveBase env ==================");
  }

  public static void clean() {
    System.out.println("================== clean HiveBase env ==================");
  }
}
```
The abstract singleton service test class:
```java
public abstract class SingletonServiceTest {
  private static boolean serviceSetUpIsDone = false;
  private static final Object lock = new Object();

  public abstract void serviceSetup();

  public abstract void serviceShutdown();

  public void setUp() {
    synchronized (lock) {
      if (!serviceSetUpIsDone) {
        System.out.println("================== setup service env ==================");
        serviceSetup();
        serviceSetUpIsDone = true;
        System.out.println("================== setup service env completed ==================");
      }
    }
  }

  public void shutdown() {
    synchronized (lock) {
      if (serviceSetUpIsDone) {
        System.out.println("================== clean service env ===================");
        serviceShutdown();
        serviceSetUpIsDone = false;
        System.out.println("================== clean service env completed ===================");
      }
    }
  }
}
```
The HMS singleton test service implementation:
```java
public class HiveMetastoreTestMain extends SingletonServiceTest {

  @Override
  public void serviceSetup() {
    HiveBase.setup();
  }

  @Override
  public void serviceShutdown() {
    HiveBase.clean();
  }
}
```
There are two different test Cases, about hive read and write UTs:
```java
public class HiveReadTest {

  private final static SingletonServiceTest HMS = new HiveMetastoreTestMain();

  @BeforeClass
  public static void test() {
    HMS.setUp();
  }

  @Test
  public void test1() {
    System.out.println("hive read test1");
  }

  @Test
  public void test2() {
    System.out.println("hive read test2");
  }
}
```
```java
public class HiveWriteTest {
  private final static SingletonServiceTest HMS = new HiveMetastoreTestMain();

  @BeforeClass
  public static void test() {
    HMS.setUp();
  }

  @Test
  public void test1() {
    System.out.println("hive write test1");
  }

  @Test
  public void test2() {
    System.out.println("hive write test2");
  }
}
```
Run the two test cases(HiveReadTest, HiveWriteTest) together and get the below result:
```
================== setup service env ==================
================== setup HiveBase env ==================
================== setup service env completed ==================
hive write test1
hive write test2
hive read test1
hive read test2
```

**Review by wangtaohz — APPROVED** [review]

**Pull request merged**